Add new integration with Opik Tracking tool

Lothiraldan opened the pull request:
# Summary

This PR integrates Opik (https://github.com/comet-ml/opik), an open-source platform for evaluating, testing, and monitoring LLM applications. It features built-in evaluation metrics and data annotation. Closes #11407, requested by a community user.

# Screenshots

| Before | After |
|--------|-------|
| ![main](https://github.com/langgenius/dify-docs/blob/main/en/.gitbook/assets/tracing-app-performance.png?raw=true) | ![branch](https://github.com/langgenius/dify-docs/blob/d440354378286d064c3fe1332c9009435d01f79e/en/.gitbook/assets/tracing-app-performance.png?raw=true)   |


# Checklist

> [!IMPORTANT]  
> Please review the checklist below before submitting your pull request.

- [x] This change requires a documentation update, included: [[Dify Document](https://github.com/langgenius/dify-docs)](https://github.com/langgenius/dify-docs/pulls)
- [x] I understand that this PR may be closed in case there was no previous discussion or issues. (This doesn't apply to typos!)
- [x] I've added a test for each change that was introduced, and I tried as much as possible to make a single atomic change.
- [x] I've updated the documentation accordingly.
- [x] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods

Lothiraldan committed:
Add new integration with Opik Tracking tool

Lothiraldan committed:
Lint fixes

Lothiraldan committed:
Add new dependency at the right place

ZhouhaoJiang commented:
![CleanShot 2024-12-12 at 15 43 28@2x](https://github.com/user-attachments/assets/58504d38-5142-4580-b3d5-ba6404588d34)
The elements have the position problem.

ZhouhaoJiang commented:
![CleanShot 2024-12-12 at 15 49 36@2x](https://github.com/user-attachments/assets/0a44b257-ff1a-45f6-82c4-a6165733349e)
![CleanShot 2024-12-12 at 15 50 49@2x](https://github.com/user-attachments/assets/bb3a100f-a212-47d6-8eb4-a793550ece5e)
When I add a credential, I don't fill in the Url field, because I see that it has a default value, but the actual request interface does not include this value, in order to better user experience should set it to the default value

ZhouhaoJiang commented:
I have tested basic chat, agent chat, and workflow and have no problems with data tracing.

Lothiraldan committed:
Add validation and proper default for Opik URL

Also add better documentation link for Opik configuration in the UI

Lothiraldan commented:
@ZhouhaoJiang Thanks for your review. I fixed the issue for the default url, it was actually on the backend. I'm guessing that the frontend was sending an empty string for it.

I'm not sure to understand your feedback about the position of the Opik Tracing Provider. I added Opik at the end of both lists, the first one with configured tracing providers and the second one with non-configured tracing providers. Do you prefer a different behavior?

ZhouhaoJiang commented:
> @ZhouhaoJiang Thanks for your review. I fixed the issue for the default url, it was actually on the backend. I'm guessing that the frontend was sending an empty string for it.
> 
> I'm not sure to understand your feedback about the position of the Opik Tracing Provider. I added Opik at the end of both lists, the first one with configured tracing providers and the second one with non-configured tracing providers. Do you prefer a different behavior?

I think my explanation is not clear. What I mean is that after one provider is configured, there is no gap between the components of the other two providers, and there is a problem with the ui

Lothiraldan committed:
Fix spacing between providers in config popup

Lothiraldan commented:
@ZhouhaoJiang I was able to fix the spacing issue but I realized that the Opik provider is missing the "IN USE" tag, I will check it out and fix the merge conflict on `api/poetry.lock`

![Screenshot 2024-12-13 at 10-30-36 Long Story Generator (Iteration) - Dify](https://github.com/user-attachments/assets/01216eb6-e402-4731-b81f-b64a085c72f9)
![Screenshot 2024-12-13 at 10-30-19 Long Story Generator (Iteration) - Dify](https://github.com/user-attachments/assets/c96fec28-50f5-4215-b5b2-a5f5033f2a10)

iamjoel commented on web/app/(commonLayout)/app/(appDetailLayout)/[appId]/overview/tracing/config-popup.tsx:
```diff
@@ -102,9 +105,62 @@ const ConfigPopup: FC<PopupProps> = ({
+    if (opikConfig)
+      configuredPanels.push(opikPanel)
+
+    console.log(configuredPanels)
```
Please remove console.log

ZhouhaoJiang reviewed (changes requested):
We have updated the workflow run id as workflow trace id and move the workflow app log id into metadata, So please update the workflow trace.
```
   def workflow_trace(self, trace_info: WorkflowTraceInfo):
        dify_trace_id = trace_info.message_id or trace_info.workflow_app_log_id or trace_info.workflow_run_id
        opik_trace_id = uuid4_to_uuid7(trace_info.start_time, dify_trace_id)

        if trace_info.message_id:
            trace_data = {
                "id": opik_trace_id,
                "name": TraceTaskName.MESSAGE_TRACE.value,
                "start_time": trace_info.start_time,
                "end_time": trace_info.end_time,
                "metadata": wrap_metadata(trace_info.metadata, message_id=trace_info.message_id),
                "input": wrap_dict("input", trace_info.workflow_run_inputs),
                "output": wrap_dict("output", trace_info.workflow_run_outputs),
                "tags": ["message", "workflow"],
                "project_name": self.project,
            }
            self.add_trace(trace_data)
```

You can refer to this pr #11585

Lothiraldan committed:
Remove console.log line

Lothiraldan commented:
@ZhouhaoJiang Thanks, I will update the workflow trace method accordingly. Can you remind me in which case the message_if field in the WorkflowTraceInfo could be None? I want to double-check that everything works nicely after the update.

Lothiraldan committed:
Use workflow_run_id as workflow trace ID or message_id if present

Lothiraldan commented:
@ZhouhaoJiang I've updated the Opik tracer to use `workflow_run_id` or `message_id` when present.

laipz8200 commented:
Thank you for this work, we'll test this feature in our development environment and may need to do some work on documentation.

Lothiraldan commented:
@laojianzi Let me know if there is anything I can do on the documentation or the code. Do you want me to take care of the poetry.lock conflict?

crazywoola commented:
@Lothiraldan Can you rebase the origin/main branch and push it again. I will merge it later.

Lothiraldan commented:
@crazywoola Sure, I rebased and pushed

crazywoola commented:
Hello, it seems that the tests are failed.

Please add this to this file https://github.com/langgenius/dify/blob/d91960905971316ef27d7da5777cfa71ac85e521/web/i18n/zh-Hans/app.ts as well 

```
 opik: {
      title: 'Opik',
      description: '一个全方位的开发者平台，适用于 LLM 驱动应用程序生命周期的每个步骤。',
 },
```

Lothiraldan commented:
@crazywoola I see it, I'm working on resolving it now. Will also update this file.

Lothiraldan committed:
Fix type hints issues in Opik Trace

Lothiraldan committed:
Merge branch 'main' into opik-monitoring

Lothiraldan committed:
Fix last type hint issues

Lothiraldan committed:
Add Opik translation

Lothiraldan committed:
Apply ruff format in addition to ruff check

Lothiraldan committed:
Update minimum Opik version to make mypy happy
